### ISSUE TITLE
Python rules do not support single quotes or spaces in path

### ISSUE THREAD

**Issue by Warggr** [snakemake/snakemake#2981]
**Snakemake version**
8.16.0

**Describe the bug**
When the path contains either single quotes `'` or spaces, Python rules (i.e. `run:`) fail.

**Logs**
The error message is
```
/bin/sh: -c: line 1: unexpected EOF while looking for matching `''
```
when the path contains a single quote and when the path contains a space, e.g. `/tmp/users\ directory`:
```
Error: Snakefile "/tmp/users" not found.
```

**Minimal example**
```sh
mkdir "user's folder" && cd user\'s\ folder
cat > Snakefile <<EOF
rule:
    run:
        print("hello, world")
EOF
snakemake
```
Output:
```
Assuming unrestricted shared filesystem usage.
Building DAG of jobs...
Using shell: /usr/bin/bash
Provided cores: 4
Rules claiming more threads will be scaled down.
Job stats:
job      count
-----  -------
1            1
total        1

Select jobs to execute...
Execute 1 jobs...

[Mon Jul 22 10:30:59 2024]
localrule 1:
    jobid: 0
    reason: Rules with neither input nor output files are always executed.
    resources: tmpdir=/tmp

/bin/sh: -c: line 1: unexpected EOF while looking for matching `''
[Mon Jul 22 10:30:59 2024]
Error in rule 1:
    jobid: 0

Shutting down, this might take some time.
Exiting because a job execution failed. Look above for error message
Complete log: .snakemake/log/2024-07-22T103059.309900.snakemake.log
WorkflowError:
At least one job did not complete successfully.
```

**Additional context**
- shell rules (i.e. `shell:` instead of `run:`) work - so it's possible to just run `shell: "python -c '...'"`

**Comment by cmeesters** [comment]
Your minimal example works like a charm for me. Are you trying to trigger Snakemake in a more elaborate shell script? Could it be that you are dealing with nested quotes in a shell script?

**Comment by Warggr** [comment]
Thanks for looking into this.

I am not using a shell script, just executing `snakemake` from the terminal.
My default shell is zsh; I thought that could maybe be the problem. But when I try running `snakemake` from Bash, or with `SHELL=bash`, or both, the same problem occurs. In any case, the error message says something about `/bin/sh`, which for me is a symbolic link to `/bin/bash`.

To make sure that the problem was not my environment, I tried running the minimal example in a Docker container, with the following Dockerfile:
```Dockerfile
FROM alpine:3.20.2

RUN apk add --no-cache py3-pip gcc python3-dev musl-dev linux-headers
RUN pip install --break-system-packages snakemake
RUN mkdir "/tmp/user's folder"
WORKDIR "/tmp/user's folder"

RUN printf "rule:\n  run:\n    print('hello, world!')" > Snakefile

CMD ["snakemake"]
```
This results in a similar error
```
/bin/sh: syntax error: unterminated quoted string
```

**Comment by Warggr** [comment]
I checked which command-line arguments `/bin/sh` is receiving, and they looked like this: 
$0: `/bin/sh`
$1: `-c`
$2: `cd '/tmp/user'"'"'s folder' && /usr/bin/python -m snakemake --snakefile /tmp/user's folder/Snakefile --target-jobs '1:' ...`
The part with `cd` seems correctly escaped, but not the argument to `--snakefile`; I think that is the cause of the problem.

**Comment by cmeesters** [comment]
yep, this is an issue in your script, not a bug in Snakemake. Your $2 line is an unterminated line. In the second part after && you have an unescaped tick mark. Executing the Snakefile you generate with the HereDoc, works.

**Comment by Warggr** [comment]
I don't have a script; I'm just calling `snakemake` on the command line, and somewhere Snakemake calls `/bin/sh`. Apparently this happens on the line https://github.com/snakemake/snakemake/blob/e8735c1477a2a82110757ba86bbd1ccbcaf327ba/snakemake/executors/local.py#L232

To investigate the problem, I replaced `/bin/sh` with a wrapper script to get the arguments that it was receiving, and printed those arguments in the previous comment. But those arguments are generated by Snakemake, not by me or by a script from me.

What do you mean by executing the Snakefile? I'm running `snakemake` in the directory where the Snakefile is; is that what you mean or is there another way?

**Comment by Warggr** [comment]
The line that generates these arguments is here: https://github.com/snakemake/snakemake-interface-executor-plugins/blob/main/snakemake_interface_executor_plugins/executors/real.py#L168

I think the problem might be the function `format_cli_arg` that is called there. Running this
```python
>>> from snakemake_interface_executor_plugins.utils import format_cli_arg
>>> print(format_cli_arg("--snakefile", "/tmp/user's folder"))
--snakefile /tmp/user's folder
```
prints an unescaped value, which is not what what I would expect.

So, should I open a bug in https://github.com/snakemake/snakemake-interface-executor-plugins instead?

**Comment by cmeesters** [comment]
I do not get it. You start with a Snakefile like this:

```Python
rule:
    run:
        print("hello, world")
```

which you can execute with `snakemake -c1` just fine, regardless of double or single quotes around the string.

Next, you present a Dockerfile with a broken shell line and subsequently an excerpt with a broken shell command to produce the mentioned Snakefile. The Snakefile is fine, the shell lines aren't.

So, which is your minimal example to produce the error you observe, if it is not the example in this thread? And why do you think there is a bug in the executor interface? You did not mention that you are using any plugin before. Btw. the link you posted is the template, not a specific plugin.

**Comment by Warggr** [comment]
Okay, thanks. Yes, the minimal example *is* the example in this thread, in my first message. Except it seems to be a system-specific bug, because the example fails on my machine but works on yours. I'm not using any plugins, just Snakemake in its default configuration.

I could reproduce it in Google Colab ([link](https://colab.research.google.com/drive/1Lm4aft2wMcjKUkeeGh7SGda_T90lXCxf?usp=sharing)), can you maybe look at that?

I've also tested the same example on a few other systems:
- Alpine Linux inside a Docker container with Python 3.12 and Snakemake 8.16: bug occurs (that was the Dockerfile I sent earlier; the Docker container builds normally, but running Snakemake inside it fails.)
- Ubuntu computer with Python 3.11 / Snakemake 8.16: bug occurs
- Windows 11 with Python 3.12 / Snakemake 8.16: works fine
- Ubuntu with Python 3.10 / Snakemake 7.32: works fine

The rest of my comments (everything related to /bin/sh, or linking to lines of code) was me trying to find the root cause of the error. Please ignore it for now.

**Comment by Warggr** [comment]
This is fixed in Snakemake 9.6.0.
